### DISCUSSION TITLE
keyspace event subscription is no working for a cluster

### DISCUSSION THREAD

**Discussion by jigneshmpatel** [redis/lettuce#2856] [Q&A]
I am trying to subscribe to the keyspace events from dockerize bitnami Redis Cluster using lettuce 6.3.2 and lettuce 4.1.109.
This is my code

```java
 public void subscribeToEncounterCheckoutTimeChannel(){
       logger.info("Inside subscribeToEncounterCheckoutTimeChannel()");
       StatefulRedisClusterPubSubConnection<String, String> pubsubConnection = connectToRedisCluster.getClusterPubSubConnection();
       pubsubConnection.setNodeMessagePropagation(true);
       //connection.addListener(new EncounterCheckoutTimeListener());
       pubsubConnection.addListener(new RedisClusterPubSubAdapter(){

           public void message(String pattern, String channel, String message) {
               String msg = message;
               logger.info("msg" + msg);

           }
       });

        //async setup
       RedisClusterPubSubAsyncCommands<String, String> asyncCommands = pubsubConnection.async();
       PubSubAsyncNodeSelection<String, String> masters =asyncCommands.upstream();
       NodeSelectionPubSubAsyncCommands<String, String> commands = masters.commands();

       logger.info("number of master nodes: "+masters.size());
       commands.psubscribe("__keyspace@0__:"+CacheConstants.EncounterCheckoutTime+"*");
         //OR
         //sync setup
       RedisClusterPubSubCommands<String, String> syncCommands = pubsubConnection.sync();

       syncCommands.upstream().commands().psubscribe("__keyspace@0__:"+CacheConstants.EncounterCheckoutTime+"*");


    }
```

But it is not able to fetch the message.  I have tried very hard to make this work but this is not working.  Is there a bug in the versions of lettuce I am using to make subscribing keyspace events? 
I have posted the information on[ Stackoverflow](https://stackoverflow.com/questions/78480322/redis-cluster-subscriber-is-not-working-with-lettuce) also. It will be really helpful if there is appropriate guidance provided to debug an make this work.

**Comment by jigneshmpatel** [comment]
The explantation from lettuce documentation is not applicable i.e. the documentation requires an update with a working solution.
https://github.com/redis/lettuce/wiki/Pub-Sub
@mp911de is there anything I refer to figure out about fixing this issue?

**Reply by mp911de** [reply]
There's nothing to fix. If you're using `RedisClusterPubSubAdapter`, then you must override the methods that provide `RedisClusterNode`. Your variant `public void message(String pattern, String channel, String message)` doesn't override any existing listener nor adapter method.

**Comment by jigneshmpatel** [comment]
Thank you very much @mp911de et all. It fixed the issue.
However seems it is not consistently getting messages(Sometimes it works) if I set a message outside Redis cluster
```
redis-cli -c config set notify-keyspace-events  Kx 
redis-cli -c SETEX EncounterCheckoutTime:2 30 "Hello, Redis
```
It doesn't print anything I believe when the original node changes, despite setting
` pubsubConnection.setNodeMessagePropagation(true);`

On a different note, I am not getting the value for the message, my listener only prints
`Got  message expired on channel __keyspace@0__:EncounterCheckoutTime:3`
But I am interested in getting key value both.

**Reply by jigneshmpatel** [reply]
To be more clear, the following works as somehow when deployed it nows about node with  IP: **172.18.0.4**
```
127.0.0.1:6379> SETEX EncounterCheckoutTime:1 120 "Hello, Redis"
-> Redirected to slot [11230] located at 172.18.0.4:6379
OK
```
but following key expiration which has been set on IP **172.18.0.5** is not received by the application.
```
172.18.0.4:6379> SETEX EncounterCheckoutTime:2 30  "Hello, Redis"
-> Redirected to slot [7101] located at  172.18.0.5:6379
```
These are cluster nodes

```
6abf93ea9d9448e7128c049e07fee65466b37c68 172.18.0.8:6379@16379 myself,master - 0 1715994847000 1 connected 0-5460
40202528f94e6929dccfffa8baafc546da4c52cb 172.18.0.2:6379@16379 slave b5a1c07679a393a13acaabebb3dac74510c906eb 0 1715994850174 7 connected
8720fcd760d54e9cb520af8ece17e0bf81b3630d 172.18.0.5:6379@16379 master - 0 1715994847079 2 connected 5461-10922
f14eee40833872d196c947cb14450b85f50374cf 172.18.0.6:6379@16379 slave 8720fcd760d54e9cb520af8ece17e0bf81b3630d 0 1715994848000 2 connected
645c7d56c16e0a73f113400c43619836c927a0c6 172.18.0.3:6379@16379 slave 6abf93ea9d9448e7128c049e07fee65466b37c68 0 1715994845013 1 connected
b5a1c07679a393a13acaabebb3dac74510c906eb 172.18.0.4:6379@16379 master - 0 1715994849126 7 connected 10923-16383
```

However, it worked when I set the following on each master node
`config set notify-keyspace-events Kx or KExed`
And after that even cluster level command started publishing too. e.g. following was not working previously
`redis-cli -c SETEX EncounterCheckoutTime:2 30 "Hello, Redis`

**Reply by jigneshmpatel** [reply]
So the missing part is what should I subscribe to instead of following:
` psubscribe("__keyspace@0__:"+CacheConstants.EncounterCheckoutTime+"*");`
so that I can receive **key id** and **value** both for expired keys.

**Reply by jigneshmpatel** [reply]
Also if the application has three servers how would it be possible to make that only one instance listen to key expiry events to avoid message duplication?

**Comment by jigneshmpatel** [comment]
Also wondering why expire message received twice
(lettuce-nioEventLoop-8-9) Got  message expired on channel __keyspace@0__:EncounterCheckoutTime:3
(lettuce-nioEventLoop-8-9) Got  message expired on channel __keyspace@0__:EncounterCheckoutTime:3

for the following command
`config set notify-keyspace-events KExed`
